### PR TITLE
http://tickets.opscode.com/browse/CHEF-2381

### PR DESCRIPTION
**Pull request by dennisklein**
http://tickets.opscode.com/browse/CHEF-2381

### CODE REVIEW

**Comment by smurawski** [comment]
@chef/maintainers Dennis Klein would love to get this merged and is willing to do what is needed to make that happen.

**Comment by danielsdeleo** [comment]
In order for this to work end to end, someone needs to make the corresponding updates to https://github.com/chef/chef_authn (and possibly https://github.com/chef/oc_erchef/blob/70c3cdd74fc2b1a7f9301d14a5edd4936aaa9ef1/apps/oc_chef_wm/src/oc_chef_wm_base.erl#L823-L858). That doesn't necessarily have to be Dennis. I'll copy my relevant review comments from the ticket to here.

File: lib/mixlib/authentication/signedheaderauth.rb [hunk 1]
```diff
@@ -79,28 +80,60 @@ def proto_version
...
         }
-
         string_to_sign = canonicalize_request(sign_algorithm, sign_version)
-        signature = Base64.encode64(private_key.private_encrypt(string_to_sign)).chomp
+        Mixlib::Authentication::Log.debug "String to sign: '#{string_to_sign}'\nHeader hash: #{header_hash.inspect}"
+        
+        # Sign
+        if sign_version <= '1.1'
+          signature = Base64.encode64(keypair.private_encrypt(string_to_sign)).chomp
+        elsif sign_version == '1.2'
```
Comment (danielsdeleo) [inline lib/mixlib/authentication/signedheaderauth.rb]: Per my comments on the ticket, I'd like to see the logic for each signing version moved to an individual class so we don't have to repeat this case logic for each rev of the algorithm, instead you pick the version up front and then tell the object for that version to give you the signature (or give you an error if you're trying to do something it doesn't support, like use signing version 1.2 with a signing algorithm other than SHA1. It's not necessary to rewrite everything from scratch, the goal is just to remove repeated cases to make the code more declarative and readable.

**Comment by dennisklein** [comment]
Thank you for your response. I will look into it.

Regarding the server part: we are using a chef server 12.0.1 currently and it works. Some erchef developer did the server-side part already a long time ago (see jira history). With current versions of knife (12+), I don't even have to patch chef any more, since the cookbook uploader was refactored and honors the `authentication_protocol_version` config parameter now.

**Comment by danielsdeleo** [comment]
I didn't realize that, that's awesome.

**Comment by thommay** [comment]
We've kind of skipped over 1.2 and are standardising on 1.3 - it's in both client and server - so I'm going to close this PR. Sorry this never got merged, @dennisklein , and thanks for your contribution.

**Comment by dennisklein** [comment]
@thommay How are we supposed to handle passphrase-encrypted private keys in version 1.3? Do you support caching the passphrase somehow now? Or do we have to type the passphrase multiple times in case of a knife cookbook upload e.g.?

**Comment by dennisklein** [comment]
@thommay Or do we skip this solution here, because there is no ssh-agent on some proprietary platform you are supporting now? Should I open a feature request for a chef-agent... :(

Please give us the real reason for not merging this really old contribution. It was implemented on server side already many years ago. 1.3 is not even a successor of 1.2? We don't care about the implementation, rewrite it, if you want, we care about the features of 1.2. What is everyone else doing? Unencrypted secrets?

**Comment by thommay** [comment]
Hi,
I understand it's upsetting when a PR is closed. But it's been untouched in over a year, and we've since bumped protocol version. There's no commercial interest in closing this PR. 
Looking at the code here, it doesn't seem unreasonable that it could be ported over to 1.3, since it  maintains a signing based approach rather than using a private key to encrypt the header. 
If you get that done, we'll be happy to take that PR.
-Thom

**Comment by dennisklein** [comment]
> since it maintains a signing based approach rather than using a private key to encrypt the header.

What does this even mean? Of course this library is "using a private key to encrypt".. that is the very definition of a cryptographic signature.

Many years back with ancient chef versions, people already raised the issue, that knife is unusable with an encrypted credential. Please refer to http://tickets.opscode.com/browse/CHEF-2381. If I recall correctly the reporter was even one of the core developers back then? I even remember this issue popping up several times. I proposed a solution by interfacing to ssh-agent and provided two different implementations. I am not interested in credit or authorship of this code, please take the code as public domain for inspiration and refactor it. The proposal was even approved already by some chef core reviewing board long ago. Many people since then have expressed interest in this. One of the chef-server developers integrated the needed server-side when chef switched to the erlang based server. Somehow - although there was only support from many chef team members over the years - this pull request got ignored...

Let us put this pull request aside and let me argue for the importance of this issue. Access to a chef server equals to root access to your infrastructure. Hence, your chef credentials should not be stored in plain text on your disk. When you encrypt your credentials, on each request to the server, knife interactively used to ask for your passphrase. This was especially unusable for a knife cookbook upload, because knife uploads each file in its own request. The most simple fix we could come up with, was to cache the passphrase for the whole knife execution.

Looking aside, we realized, that ssh already solved this issue (pgp and many other tools, too) by providing the ssh-agent. So, we introduced authentication protocol version 1.2 to be able to use signatures generated by the ssh-agent. Please read the history on http://tickets.opscode.com/browse/CHEF-2381 for all the details. Using the ssh-agent also brings the benefit of being able to use all kinds of security tokens already supported by ssh - we have successfully used multi-factor credential stores like smartcards to sign chef server request using this patch.

> If you get that done, we'll be happy to take that PR.

Question is, has chef solved this issue nowadays otherwise? Not to my knowledge, but I am not up to date with master. I personally will definitely not put in the work to rebase this pull request on top of 1.3. To what end? That you ignore it for the next three years?

By rejecting this feature after this long time and in case you have not solved the issue with plain text chef credentials otherwise the chef project strongly states it does not care about thousands of plain text keys on probably thousands of laptops which give root access to whole infrastructures. :+1:

**Comment by btm** [comment]
@dennisklein, you care about this feature, believe it is an essential security feature, and are frustrated that it didn't make it to getting merged. I'm sorry the process failed you. 

The Chef maintainers have tried streamlining the contribution process to make it easier to contribute over the years, such consolidating onto GitHub rather than using both GitHub and JIRA which was likely related to your problems using tickets.opscode.com. The http://github.com/chef/chef-rfc process is a new way to get community agreement on a design to streamline it's acceptance. Given the number of contributions, it's difficult for the maintainers to stay engaged in a two-way conversation on an issue that has a turn-around time of many days between feedback and responses. The side-projects in particular are easy for us to miss.

The feature isn't rejected, but this pull request hasn't been moving forward. Pruning old and stalled pull requests helps us identify and engage active pull requests. I'm not aware of existing support for encrypted private keys in the latest versions of Chef. I suspect it hasn't been a big enough issue for existing users.

There's interest in having this feature as you've indicated, but there hasn't been enough interest by anyone for them to implement it. Interest alone isn't going to make the chef maintainers finish it for you. We will do our best to provide feedback to contributions to get them included in the project, but can't be expected to finish pull requests for contributors as wouldn't have time to review all the other contributions, let alone for our jobs.

Again, I'm sorry that you're so frustrated that you don't want to finish your contribution. Hopefully another contributor will take interest, or someone without the time or ability to contribute will sponsor a company to do it.

**Comment by dennisklein** [comment]
> There's interest in having this feature as you've indicated, but there **hasn't been enough interest by anyone for them to implement it**

?? You got two production ready implementations by me one merge button click away which are still used by 10+ engineers in our institute up to today without issues.

> but this pull request hasn't been moving forward. Pruning old and stalled pull requests helps us identify and engage active pull requests.

You have released version 1.3.0 of this library in August 2012, the pull request was from March 2013. There was no development until late 2015 when preparation for 1.4.0 was started. So, the maintainers of this library had roughly 29 months for clicking a button and maybe adjusting the field with the version number for a new release. :+1:

**Comment by btm** [comment]
This pull request was never considered complete, and it's less complete now due to merge conflicts. If you'd like to move forward with this feature, get it rebased and the maintainers will re-review it. There's no other way forward. 

You've stated you will not finish this work, so this feature is dead until you change your mind or another contributor becomes interested in completing it.

**Comment by chuhn** [comment]
We are relying on this patch to manage ~2000 nodes in our institute. I'd consider it "complete" as it just works.
One of our best developers put some serious work into developing this patch, signed your CLA and did everything necessary to get this patch forward.

Within 3 years you apparently did nothing. I suspect that it would be straightforward for @dennisklein to upgrade the patch to the current version of mixlib-auth, but given your engagement, we will be in the same situation again in 2019?

What alternatives to this patch can you offer?
AFAICS if the admin keys are encrypted you have to enter your password on every knife invocation (actually multiple times last time I tried).
If the admin keys are unencrypted we are solely relying on file-system security which is totally unsuitable for credentials that basically grants root access to an infrastructure of multiple thousand nodes.

**Comment by thommay** [comment]
Hi @chuhn, @dennisklein ;
we appreciate that it's upsetting when PRs get closed. However, in this case, several Chef employees spent time and effort reviewing this PR. They requested some changes. At that point Dennis said that he would look in to those changes ( https://github.com/chef/mixlib-authentication/pull/4#issuecomment-82581813 ), and then nothing happened for a year.
Chef is a small company; we rely on external contributions for features that we don't have the resources to build ourselves, and we have a workflow for managing those contributions. When we don't hear anything from the contributor for a year, we have to assume they've lost interest or have moved on to other things.
As we've said, if this PR is updated to version 1.3 of the protocol, we'd be happy to review and take that contribution.
Thanks,
-Thom

**Comment by chuhn** [comment]
Frankly I'm rather disappointed than upset.

> Chef is a small company; we rely on external contributions for features that we don't have the resources to build ourselves

This is what we tried to provide. Your response was not so encouraging. Or did I miss the irony here as a non-native speaker?

To get back to https://tickets.opscode.com/browse/CHEF-2381 from **May 2011**:
> Knife should be able to deal with encrypted private keys.
> The reason: the same reason that we encrypt our SSH private keys.

This issue is still unresolved. Obviously this would be an important security measure. 
How does Facebook deal with that? Etsy?

**Comment by thommay** [comment]
Hi,
there was no irony - but you're missing the key point. This PR is unmerged because the original submitter chose not to finish it. Yes, Chef could conceivably have done the work to make the patch production quality and maintainable, but in the absence of anyone else asking us to do this, we've prioritised other features.

**Pull request closed without merging**